NDTensors/src/lib/BlockSparseArrays/src/blocksparsearray/blockdiagonalarray.jl

```diff
@@ -0,0 +1,60 @@
+# type alias for block-diagonal
+using LinearAlgebra: Diagonal
+
+const BlockDiagonal{T,A,Axes,V<:AbstractVector{A}} = BlockSparseMatrix{
+  T,A,Diagonal{A,V},Axes
+}
+
+function BlockDiagonal(blocks::AbstractVector{<:AbstractMatrix})
+  return BlockSparseArray(
+    Diagonal(blocks), (blockedrange(size.(blocks, 1)), blockedrange(size.(blocks, 2)))
+  )
+end
+
+# Cast to block-diagonal implementation if permuted-blockdiagonal
+function find_permute_blockdiagonal(A)
+  @show keys = map(Base.Fix2(getproperty, :n), collect(block_stored_indices(A)))
+  I = first.(keys)
+  allunique(I) || return nothing
+
+  J = last.(keys)
+  p = sortperm(J)
+  Jsorted = J[p]
+  allunique(Jsorted) || return nothing
+
+  return I[p], Jsorted
+end
+
+"""
+    try_make_blockdiagonal(A)
+
+Attempt to find a permutation of blocks that makes `A` blockdiagonal. If unsuccesful,
+returns nothing, otherwise returns both the blockdiagonal `B` as well as the permutation `I, J`.
+"""
+function try_make_blockdiagonal(A::AbstractBlockSparseMatrix)
```

Comment:
Consider changing to `try_to_blockdiagonal` as discussed above.

Reply:
Also, does this handle rectangular block matrices, i.e. when the block pattern itself is rectangular?